src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java

```diff
@@ -254,4 +257,107 @@ public void close() {
+   *
+   * @return a String value contains the offset token
+   */
+  public String getClientStatus() {
+    ClientStatusResponse response;
+    try {
+      response =
+          (ClientStatusResponse)
+              InternalUtils.backoffAndRetry(
+                  telemetry,
+                  SnowflakeInternalOperations.GET_CLIENT_STATUS_SNOWPIPE_API,
+                  () -> ingestManager.getClientStatus(null));
+    } catch (Exception e) {
+      throw SnowflakeErrors.ERROR_3007.getException(e);
+    }
+    if (response != null) {
+      // offsetToken is ok to be null
+      return response.getOffsetToken();
+    } else {
+      throw SnowflakeErrors.ERROR_4001.getException("the response of get client status is null");
+    }
+  }
+
+  /**
+   * Ingest a list of files with the clientInfo (clientSequencer and offsetToken)
+   *
+   * @param fileNames file name List
+   * @param clientSequencer unique identification of the Snowpipe client
+   */
+  public void ingestFilesWithClientInfo(List<String> fileNames, long clientSequencer) {
+    if (fileNames.isEmpty()) {
+      return;
+    }
+    logDebug("ingest files with client info: {}, clientSequencer: {} ", fileNames, clientSequencer);
+    try {
+      InternalUtils.backoffAndRetry(
+          telemetry,
+          SnowflakeInternalOperations.INSERT_FILES_WITH_CLIENT_INFO_SNOWPIPE_API,
+          () -> {
+            while (fileNames.size() > 0) {
+              // Can not send more than 5000 files in one request,
+              // so batch 4000 as one request
+              int toIndex = Math.min(4000, fileNames.size());
+              List<String> fileNamesBatch = fileNames.subList(0, toIndex);
+              String offsetToken = getLastOffsetTokenFromBatch(fileNamesBatch);
+              InsertFilesClientInfo clientInfo =
+                  new InsertFilesClientInfo(clientSequencer, offsetToken);
+              Set<String> fileNamesSet = new HashSet<>(fileNamesBatch);
+              ingestManager.ingestFiles(
+                  SimpleIngestManager.wrapFilepaths(fileNamesSet), null, false, clientInfo);
+              fileNamesBatch.clear();
+            }
+            return true;
+          });
+    } catch (Exception e) {
+      throw SnowflakeErrors.ERROR_3008.getException(e);
+    }
+  }
+
+  /**
+   * Get the last offset number from a list of fileName
+   *
+   * @param fileNameBatch
+   * @return Offset number in String format
+   */
+  private String getLastOffsetTokenFromBatch(List<String> fileNameBatch) {
+    Long lastFileEndOffset =
+        FileNameUtils.fileNameToEndOffset(fileNameBatch.get(fileNameBatch.size() - 1));
+    for (String fileName : fileNameBatch) {
+      if (lastFileEndOffset < FileNameUtils.fileNameToEndOffset(fileName)) {
+        lastFileEndOffset = FileNameUtils.fileNameToEndOffset(fileName);
+        logWarn("The fileName list is not sequential.");
+      }
+    }
+    return lastFileEndOffset.toString();
```

Comment:
can you add a unit test for this? You can make this protected and add annotation @visiblefortesting

Reply:
or you can add multiple files in existing test in different order and see if it works.